openvdb/math/Stencils.h

```diff
@@ -1701,6 +1702,119 @@ class DenseStencil: public BaseStencil<DenseStencil<GridT, IsSafe>, GridT, IsSaf
 
+//////////////////////////////////////////////////////////////////////
+
+
+/// @brief Dense stencil representing a convolution kernel
+/// @note The source grid is assumed to be unchanging.
+template<typename GridT, bool IsSafe = true>
+class ConvolutionStencil: public BaseStencil<ConvolutionStencil<GridT, IsSafe>, GridT, IsSafe>
+{
+public:
+    using SelfT = ConvolutionStencil<GridT, IsSafe>;
+    using BaseType = BaseStencil<SelfT, GridT, IsSafe>;
+    using GridType = GridT;
+    using ValueType = typename GridType::ValueType;
+    using KernelValueType = typename std::conditional<
+        std::is_floating_point<ValueType>::value, ValueType, double>::type;
+    using KernelType = typename boost::multi_array<KernelValueType, 3>;
+    using KernelCRefType = typename boost::const_multi_array_ref<KernelValueType, 3>;
+    using StencilRefType = typename boost::multi_array_ref<ValueType, 3>;
+
+
+    ConvolutionStencil(const GridType& grid, const KernelCRefType& kernel)
+        : BaseType(grid, static_cast<int>(kernel.num_elements()))
+        , mKernel(kernel)
+        , mKernelShape(mKernel.shape(), mKernel.shape() + 3)
+    {
+    }
+
+    /// @brief Populate the stencil buffer with the values of voxel (i, j, k) and its neighbors.
+    void moveTo(const Coord& ijk) { this->init(ijk); }
+
+    /// @brief Populate the stencil buffer with the values of the voxel to which
+    /// the given iterator points and the values of that voxel's neighbors.
+    template<typename IterType>
+    void moveTo(const IterType& iter) { this->init(iter.getCoord()); }
+
+    ValueType convolve() const
+    {
+        auto* kernelPtr = mKernel.origin();
+        auto sum = zeroVal<ValueType>();
+        for (size_t n = 0; n < mStencil.size(); ++n) {
+            /// @todo Suppress type conversion warnings (when ValueType is not float or double).
+            sum = static_cast<ValueType>(sum + (mStencil[n] * kernelPtr[n]));
+        }
+        return sum;
+    }
+
+private:
+    void setCenter(const Coord& ijk) { BaseType::mCenter = ijk; mInitialized = true; }
+
+    void init(const Coord& ijk)
+    {
+        const Coord shape(
+            static_cast<Int32>(mKernelShape[0]),
+            static_cast<Int32>(mKernelShape[1]),
+            static_cast<Int32>(mKernelShape[2]));
+        const Coord offset = shape >> 1;
+
+        if (mInitialized && (ijk == this->getCenterCoord().offsetBy(0, 0, 1))) {
+            // If the stencil buffer has been initialized and its new position differs
+            // from the old by +1 along the z axis (the common case for convolution),
+            // preserve the contents of the buffer and sample only the new z slice.
+            // (Note that BaseStencil::moveTo() sets mCenter before calling init(),
+            // inhibiting this optimization.)
+            /// @todo This optimization could be applied for any offset that leaves some
+            /// overlap between the old and new stencils, but the bookkeeping gets messy.
```

Comment:
We've also run into these kind of problems in the past too, especially for larger kernel sizes - typically 4x4 for cubic b splines which cache a 5x5 region for a single voxel which may sample positions either side. It would be awesome if the stencil framework supported optimizations for trivial stepping through neighbours!

Reply:
@Idclip are you referring to incrementally modifying the stencil buffer as it is translated (as oppose to repopulating the entire buffer from scratch)? If so, this is something we've though about as well but never found an efficient implementation (the logic for checking which grid points can be reused easily outweighs the benefit of avoiding a reload - at least that's my experience).

Reply:
oh I see - this optimization is limited to the special case where the stencil is translated by one voxel along the z-axis. Assuming that's a sufficiently common case that's a great idea (and obviously much more tractable then the general case I was referring to)

Reply:
> @Idclip are you referring to incrementally modifying the stencil buffer as it is translated (as oppose to repopulating the entire buffer from scratch)? 
Correct, I'd be interested to know how much time this optimisation of handling a single +z step shaves off, as although arbitrary size steps with arbitrary width stencils could be pretty complex, +/- stepping a single voxel in any direction (or perhaps just faces) could be worth investigating for all stencils.

Reply:
Some Houdini timings for the sharpening filter both with and without the +z optimization and, for comparison, for the existing Gaussian smoothing filter:
<table>
<tr><td></td><td align=center><b>crawler</b><br><sub>181 MVox</sub></td><td align=center><b>dragon</b><br><sub>23 MVox</sub></td><td align=center><b>venus</b><br><sub>6 MVox</sub></td></tr>
<tr><td><b>5&times;5&times;5</b><br>&nbsp;&nbsp;Sharpening<br>&nbsp;&nbsp;Optimized Sharpening<br>&nbsp;&nbsp;Gaussian Smoothing</td><td><br>61.3s<br>34.3s<br>17.8s</td><td><br>8.0s<br>4.6s<br>2.8s</td><td><br>2.0s<br>1.2s<br>1.0s</td></tr>
<tr><td><b>7&times;7&times;7</b><br>&nbsp;&nbsp;Sharpening<br>&nbsp;&nbsp;Optimized Sharpening<br>&nbsp;&nbsp;Gaussian Smoothing</td><td><br>144.6s<br>76.6s<br>20.1s</td><td><br>18.6s<br>10.3s<br>3.5s</td><td><br>4.7s<br>2.6s<br>1.1s</td></tr>
<tr><td><b>9&times;9&times;9</b><br>&nbsp;&nbsp;Sharpening<br>&nbsp;&nbsp;Optimized Sharpening<br>&nbsp;&nbsp;Gaussian Smoothing</td><td><br>292.3s<br>146.3s<br>22.3s</td><td><br>37.8s<br>20.0s<br>3.5s</td><td><br>9.4s<br>5.0s<br>1.1s</td></tr>
</table>

Reply:
Nice results, thanks @pcucka! Are you able to calculate/estimate the probability of stepping in the more common case (+z) for the convolution stencil (as I'm not quite clear algorithmically why stepping in +z is more common)?  For example, 50% of all stencil moves are in +z. Or, to put it another way, I'm wondering if we can estimate the potential performance gains from introducing this type of optimisation more generally, as I think you could support -z simply by changing the inner loop to `for (int k = 0; k < shape[2] - 1; ++k)` and other face offsets with a similar tweak.

Reply:
> I'm not quite clear algorithmically why stepping in +z is more common

All of the existing filters, as well as the sharpening filter, iterate over leaf-level active values using a `ValueOnIter`, which is arguably the "natural" way to implement a filter.  `ValueOnIter` loops over leaf buffer indices (via `NodeMask::OnIterator`), and that corresponds to _xyz_ iteration with _z_ as the fastest-changing coordinate.  So that's why I chose to optimize the +_z_ case first.

The other common case, I suppose, would be _xyz_ iteration over a bounding box, but there too the _z_ loop is typically the inner loop.

Reply:
Just to clarify, the sharpening kernel is not a separable kernel, so I would expect an unsharp mask filter to significantly outperform this sharpening filter.  (That is what I was trying to suggest by including timings for the Gaussian filter.)  That said, in none of my experiments with unsharp masking was I able to produce surfaces without holes.